chain/syncer.go

```diff
@@ -421,13 +345,42 @@ func (syncer *Syncer) HandleNewTipset(ctx context.Context, tsKey types.TipSetKey
 		return nil
 	}
 
-	// Walk the chain given by the input blocks back to a known tipset in
-	// the store. This is the only code that may go to the network to
-	// resolve cids to blocks.
-	chain, err := syncer.collectChain(ctx, tsKey)
-	if err != nil {
-		return err
+	var chain []types.TipSet
+	switch syncer.syncMode {
+	case Syncing:
+		chain, err = syncer.fetcher.FetchTipSets(ctx, tsKey, func(t types.TipSetKey) bool {
```

Comment:
It looks like we could deduplicate the termination function by declaring it at the top level.